### PR TITLE
Set up minikube for k8s testing

### PR DESCRIPTION
**Pull request by umohnani8**
Install and set up minikube so that we can
create a k8s cluster for testing.

Signed-off-by: Urvashi Mohnani <umohnani@redhat.com>

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

**Comment by umohnani8** [comment]
@cevich PTAL

File: contrib/cirrus/setup_environment.sh
Comment (cevich) [inline contrib/cirrus/setup_environment.sh:314]: Do you know if the above command will actually run podman in any way?  If so, you may want to move it down a line (below `install_test_configs`).

File: contrib/cirrus/runner.sh
Comment (cevich) [inline contrib/cirrus/runner.sh:139]:
```suggestion
bats test/minikube/001-kube.bats
```

I think?

**Comment by cevich** [comment]
Looks like a good start to me.  To speed up development, please feel free to do a temporary `DO NOT MERGE` commit that removes the `rootless_system_test` dependency.  For other tasks (except `build`), you can bypass them by (temporarily) adding a `skip: $CI == $CI` item to them.  That'll should get the minikube task running earlier so you can iterate faster :grin:

File: contrib/cirrus/setup_environment.sh
Comment (umohnani8) [reply contrib/cirrus/setup_environment.sh:314]: it won't run podman, it just configures minikube to use podman when we start it.

**Comment by cevich** [comment]
Oh, and you'll want to change `IMAGE_SUFFIX: "c5823947156488192"` to `IMAGE_SUFFIX: "c5167094042984448"`

File: contrib/cirrus/runner.sh
Comment (umohnani8) [reply contrib/cirrus/runner.sh:139]: @cevich can I actually remove this completely? 

I want to add the new minikube tests to https://github.com/containers/podman/blob/main/test/system/710-kube.bats so I can use the helper functions from helpers.bash.

Comment (cevich) [reply contrib/cirrus/runner.sh:139]: I have no problem with you changing whatever is needed.  Topic-wise, this is definitely a system-test, I'm just not sure about all the ramifications or requirements of including it in **those** system tests.

For example as I understand it, those tests are run in the Fedora/RHEL testing systems.  So, a very different environment entirely from Cirrus-CI, and our VMs.  So that's definitely something you'd want to talk with @edsantiago about.

A hack-ish idea could be to have it in a separate directory but symlink the helper script.  Not sure how Ed would feel about that, since it could become hard to remember the helpers have multiple usage-contexts.  Then again, any breakage should show up by a CI test failing, so maybe it's not so bad.

**Comment by cevich** [comment]
FYI: The new task is throwing `Error: Unable to find a match: /var/cache/download/minikube-latest*` because the `IMAGE_SUFFIX` in `.cirrus.yml` is wrong (see my comment above).

File: .cirrus.yml
Comment (cevich) [inline .cirrus.yml:38]:
```suggestion
    IMAGE_SUFFIX: "c5167094042984448"
```

This is causing the build-task failures

**Comment by umohnani8** [comment]
@cevich are these failures related to my changes? Did I mess up something?

File: .cirrus.yml
Comment (umohnani8) [reply .cirrus.yml]: Thanks! Fixed :)

**Comment by umohnani8** [comment]
@cevich any idea why the minikube test was cancelled?

**Comment by edsantiago** [comment]
@umohnani8 the `minikube` test depends on `rootless_system_test`, which in turn depends on `rootless_integration_test`, which failed. The failure looks real, because many tests failed in the same way. I suspect a problem with the VM images.

**Comment by cevich** [comment]
[Ahh this is the "os error 95" thing.](https://github.com/containers/netavark/issues/426)  This is caused by a bug in netavark (I think).  [There's a fix that should be in the stable repos by now.](https://bodhi.fedoraproject.org/updates/FEDORA-2022-29c1ba1101).  Getting it into CI means building new VM Images.  I've got an unrelated PR doing that now, I just need to flip some switches.  Can you wait a day or two or you need something sooner?

**Comment by umohnani8** [comment]
@cevich thanks, I can wait a day or so for it to be done.

**Comment by cevich** [comment]
Great, if that changes let me know and I'll do a separate build for this.  Otherwise they can come from https://github.com/containers/automation_images/pull/213 - if only the damn CI would pass (lol...it's my fault, I'm trying to add new tests).

**Comment by cevich** [comment]
okay @umohnani8 I think this should work.  Give `IMAGE_SUFFIX: "c4678746211876864"` a try and let me know.

**Comment by TomSweeneyRedHat** [comment]
LGTM
once the tests are hip.  Hopefully @cevich 's thought cures all ills, at least in this PR.

File: test/minikube/001-kube.bats
Comment (cevich) [inline test/minikube/001-kube.bats:11]: (I'm assuming setup() is called before every test)
Suggestion: Does minikube share state between startups?  If so, it would be helpful as a general testing strategy to give every "start" it's own unique state.  This prevents bad data from one failed test impacting the operations of the next test.

File: test/minikube/001-kube.bats
Comment (cevich) [inline test/minikube/001-kube.bats:27]: (Caveat: I'm not a bats expert)
Minor nit: in bash, `[` is normally an alias for `/bin/test` where as `[[` is a bash built-in.  It's slightly confusing to see the two being used together like this without a comment explaining why.  It would be good to choose one and stick with it.  Or, if there's a good reason for deviating, add a comment so others understand why.

File: test/minikube/001-kube.bats
Comment (cevich) [inline test/minikube/001-kube.bats:24]: Maybe it's my ignorance, but testing `*"Ready"*` seems odd.  Did you mean: `[[ "$output" =~ Ready ]]` (match the regular expression `Ready`)?

If not, checking equality `==` here may be a bit unreliable.  It's conceivable a future version of minikube outputs something slightly different, or a warning message or something causing a flake or annoying-failure.

File: test/minikube/001-kube.bats
Comment (edsantiago) [inline test/minikube/001-kube.bats:57]: Eeeeeek! Please never, ever, ever use this. Although this is normal in BATS-land, it is a nightmare in terms of usability. When it fails, all it tells you is something like "duh, failed" but without any actual useful information.

`run_podman` already does status checks. It is built-in. With a numeric arg, like `run_podman 1`, it will guarantee that exit code is 1, and fail (with an actually useful message) if it isn't.

File: test/minikube/001-kube.bats
Comment (edsantiago) [inline test/minikube/001-kube.bats:67]: Even in this use, `run` other than `run_podman`, this sort of test is horrible. A better thing to do is
```
run minikube ...
assert "$status" -eq 0 "exit status of minikube etc"
```

**Review by edsantiago — COMMENTED** [review]
Came here to agree with @cevich's assessments, found a few more concerns

File: test/minikube/001-kube.bats
Comment (cevich) [inline test/minikube/001-kube.bats:31]: Consider using a randomly generated container name *and* verifying it doesn't already exist.

File: test/minikube/001-kube.bats
Comment (cevich) [inline test/minikube/001-kube.bats:43]: Non-synchronize-able actions like `sleep` are almost always a bad idea to use in tests since there is very little control over the operating conditions at this level.  I highly recommend using some kind of timeout-wait loop to synchronize against some known/expected state.

File: test/minikube/001-kube.bats
Comment (cevich) [inline test/minikube/001-kube.bats:53]: Same random-name + check `exists` suggestion applies here.  Alternatively (for all tests) give podman (and minikube?) a unique root and run-root directory to store state.  Then nobody ever has to worry about clashes (they're super-duper annoying to debug).

**Review by cevich — COMMENTED** [review]
This is good progress.  I would really like @edsantiago to comment since he knows BATS (and bash) far better than I.

File: test/minikube/001-kube.bats
Comment (umohnani8) [reply test/minikube/001-kube.bats:93]: Fixed

File: test/minikube/001-kube.bats
Comment (umohnani8) [reply test/minikube/001-kube.bats:11]: We don't want to start a minikube cluster before every test, it takes quite a few seconds for a cluster to come up, so to delete and restart before each test will be too time consuming, especially as we add more tests. That is why, in the setup function I have a check to see if we are at the first test and to only spin up a minikube cluster then.

I can add a cleanup function that cleans up anything created during the test at the end of it so we basically have a fresh cluster for each test without having to start and delete it for every test.

File: test/minikube/001-kube.bats
Comment (umohnani8) [reply test/minikube/001-kube.bats:44]: No reason, switched all to `[[`

File: test/minikube/001-kube.bats
Comment (umohnani8) [reply test/minikube/001-kube.bats]: Fixed

File: test/minikube/001-kube.bats
Comment (umohnani8) [reply test/minikube/001-kube.bats]: mktemp here creates the file, `kube generate` fails to write to a file if it already exists. I have updated the filename here to have minikube in it, but it is a randomly generate name at the end of the day, so don't think any file clashes should happen here.

File: test/minikube/001-kube.bats
Comment (umohnani8) [reply test/minikube/001-kube.bats]: Fixed

File: test/minikube/001-kube.bats
Comment (umohnani8) [reply test/minikube/001-kube.bats]: Fixed for status, but I am not seeing a way to user `assert` to check for a substring in the output.

File: test/minikube/001-kube.bats
Comment (umohnani8) [reply test/minikube/001-kube.bats:48]: The name that comes with a randomly generated name is not supported by k8s, they have some specifications around what the name of the container/pod can be.

File: test/minikube/001-kube.bats
Comment (umohnani8) [reply test/minikube/001-kube.bats]: Fixed

File: test/minikube/001-kube.bats
Comment (umohnani8) [reply test/minikube/001-kube.bats:67]: Where/how can I change the root and run-root directories for these tests?

File: test/minikube/001-kube.bats
Comment (umohnani8) [reply test/minikube/001-kube.bats]: Fixed

File: test/minikube/001-kube.bats
Comment (edsantiago) [reply test/minikube/001-kube.bats]:
```console
$ ack ' assert .*=~' test/system
```
(or `grep -R`, or whatever your favorite search tool is)

File: test/minikube/001-kube.bats
Comment (edsantiago) [reply test/minikube/001-kube.bats:67]: Ugh - that's varsity-level. System tests were written with the specific goal of using the default system setup, so I made no provision for things like that. If you absolutely have to, ac/grep for `_PODMAN_TESTS_OPTS` and `runroot`, and see (e.g.) the 330 and 520 tests.

File: test/minikube/001-kube.bats
Comment (umohnani8) [reply test/minikube/001-kube.bats]: thanks, fixed

File: test/minikube/001-kube.bats
Comment (umohnani8) [reply test/minikube/001-kube.bats:73]: I also don't think this should be a problem because I am calling `basic_teardown()` after each tests which deletes all pods and containers before moving on to the next test.

File: test/minikube/001-kube.bats
Comment (edsantiago) [inline test/minikube/001-kube.bats:6]: Recommendation: set up your own `helpers.bash` in this subdirectory. From it, do the `load ../whatever` and the first-and-last-test thing. (Unless you plan to only ever have `001.bats`, which seems unlikely)

File: test/minikube/001-kube.bats
Comment (edsantiago) [inline test/minikube/001-kube.bats:18]: Also belongs in your own local helpers.bash

File: test/minikube/001-kube.bats
Comment (edsantiago) [inline test/minikube/001-kube.bats:93]: Also etc

File: test/minikube/001-kube.bats
Comment (cevich) [reply test/minikube/001-kube.bats:73]: That may be good enough for now.  Especially if there's a failure to remove and it logs that fact.

File: contrib/cirrus/runner.sh
Comment (edsantiago) [inline contrib/cirrus/runner.sh:124]: Remove the `001`, just run bats on the entire directory. Unless you only ever gonna have `001.bats`

**Review by edsantiago — COMMENTED** [review]
This is moving waaaaaaaaaaaaay too fast for slow old me. I will step back until things calm down (like, tomorrow morning).

File: test/minikube/001-kube.bats
Comment (cevich) [inline test/minikube/001-kube.bats:38]: This timeout/wait LGTM, thanks for adding it.

File: test/minikube/001-kube.bats
Comment (cevich) [reply test/minikube/001-kube.bats:53]: okay, gotcha. I think collecting the minikube log may be good enough coverage for any name-clash problems.

File: test/minikube/001-kube.bats
Comment (cevich) [reply test/minikube/001-kube.bats]: okay, SGTM.

File: test/minikube/001-kube.bats
Comment (edsantiago) [inline test/minikube/001-kube.bats:66]: Unnecessary (`wait_for_pods_to_start` dies on error; and it does not set `$status`). Please remove.

File: test/minikube/001-kube.bats
Comment (edsantiago) [inline test/minikube/001-kube.bats:67]: This, though, probably merits a status check

**Review by edsantiago — CHANGES_REQUESTED** [review]
Final set of comments on this iteration. All of them are nits: in broad sense, looks like good testing. Thanks for your tolerance of my nitpicking!

File: test/minikube/001-kube.bats
Comment (umohnani8) [reply test/minikube/001-kube.bats]: fixed

File: test/minikube/001-kube.bats
Comment (umohnani8) [reply test/minikube/001-kube.bats]: fixed

File: test/minikube/001-kube.bats
Comment (umohnani8) [reply test/minikube/001-kube.bats]: fixed

File: contrib/cirrus/runner.sh
Comment (umohnani8) [reply contrib/cirrus/runner.sh]: fixed

File: test/minikube/001-kube.bats
Comment (umohnani8) [reply test/minikube/001-kube.bats]: fixed

File: test/minikube/001-kube.bats
Comment (umohnani8) [reply test/minikube/001-kube.bats]: fixed

File: test/minikube/001-kube.bats
Comment (umohnani8) [reply test/minikube/001-kube.bats]: fixed

File: test/minikube/001-kube.bats
Comment (umohnani8) [reply test/minikube/001-kube.bats]: fixed

File: test/minikube/001-kube.bats
Comment (umohnani8) [reply test/minikube/001-kube.bats:36]: fixed

File: test/minikube/001-kube.bats
Comment (umohnani8) [reply test/minikube/001-kube.bats]: Can I do this in a follow up PR? I would like to get this in first so we can unblock the kube apply PR

I am also separating the tests out into different k8s namespaces, so one test shouldn't affect another if something goes wrong.

**Comment by edsantiago** [comment]
Reviewing again. The new `helpers.bash` needs a trailing newline to pass validation. Since you need to resubmit anyway, would you consider calling `basic_setup` from your `setup`? It will prefetch `$IMAGE` and do some cleanup that might give you an extra edge in running tests after one test fails.

**Comment by umohnani8** [comment]
Thanks for all the reviews @edsantiago and @cevich!

**Review by edsantiago — APPROVED** [review]

**Comment by edsantiago** [comment]
New tests [failing](https://api.cirrus-ci.com/v1/artifact/task/5151223987830784/html/minikube-podman-fedora-36-root-host.log.html):
```
[+0004s] not ok 1 minikube - check cluster is up
         # (from function `setup' in test file test/minikube/[helpers.bash, line 8](https://github.com/containers/podman/blob/4a04b156eeccf30fcfedf4175f09edbc1c407728/test/system/helpers.bash#L8))
         #   `minikube start' failed with status 57
         # * minikube v1.27.0 on Fedora 36 (kvm/amd64)
         # ! Kubernetes 1.25.0 has a known issue with resolv.conf. minikube is using a workaround that should work for most use cases.
         # ! For more information, see: https://github.com/kubernetes/kubernetes/issues/112135
         # * Using the podman driver based on user configuration
         # * The "podman" driver should not be used with root privileges. If you wish to continue as root, use --force.
         # * If you are running minikube within a VM, consider using --driver=none:
         # *   https://minikube.sigs.k8s.io/docs/reference/drivers/none/
         #
         # X Exiting due to DRV_AS_ROOT: The "podman" driver should not be used with root privileges.
```
Not something I can help with, sorry. Or at least, easily help with. Am posting in case it's something trivial for you to identify and fix. If it isn't, lmk and I'll spin up a VM.

**Comment by edsantiago** [comment]
The `sys` failure is a flake, I've filed #16075 for it but there's no point in rerunning it. Anyhow, don't sweat that one.

**Comment by umohnani8** [comment]
@cevich can minikube be run rootless? root privilege are not required for it when using podman

File: .cirrus.yml
Comment (edsantiago) [inline .cirrus.yml:770]: >can minikube be run rootless? root privilege are not required for it when using podman

@umohnani8 I interpret your question as "how can we make this CI test run rootless". If that interpretation is correct, I believe the answer is, add `PRIV_NAME: rootless` immediately after this TEST_FLAVOR one, at the same indentation. (If you're asking about minikube itself, I can't help).

File: test/minikube/001-kube.bats
Comment (cevich) [reply test/minikube/001-kube.bats]: Yep no problem, and I like your namespace idea too.

File: .cirrus.yml
Comment (cevich) [reply .cirrus.yml:770]: Correct.  Though reading through the issue a bit, it seems to center around having a FQDN hostname set.  That should **not** be the case here. The CI hostnames should all be `cirrus-task-<blahblahblah>` or `localhost`.  In any case, I'm fine either way (root or rootless).  Though rootless brings significant additional setup complexity to the environment.  That may be worthy of avoiding if the simple/suggested `--driver=none` could work?

**Comment by umohnani8** [comment]
All tests are finally green! @cevich @edsantiago can I please get a final lgtm to merge.

File: .cirrus.yml
Comment (umohnani8) [reply .cirrus.yml:784]: yup, that is exactly what I wanted and it works, thanks!

**Pull request merged**
